### PR TITLE
common : move arg parser code to `arg.cpp`

### PR DESCRIPTION
**Pull request by ngxson**
Ref comment: https://github.com/ggerganov/llama.cpp/pull/9308#discussion_r1748739611

In this PR:
- Move arg parser code to `arg.cpp` and `arg.h`
- Add `set_sparam()` to mark a specific arg as "sampling param"
- Re-categorized some args to the correct example (for example: `llama-lookup` and `llama-parallel`)

(Tested) Build time is not increased compared to master branch.

---

- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [ ] Low
  - [x] Medium
  - [ ] High

### CODE REVIEW

File: common/arg.h [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    bool has_value_from_env();
+    std::string to_string();
+};
+
+struct llama_arg_context {
+    enum llama_example ex = LLAMA_EXAMPLE_COMMON;
+    gpt_params & params;
+    std::vector<llama_arg> options;
+    void(*print_usage)(int, char **) = nullptr;
+    llama_arg_context(gpt_params & params) : params(params) {}
```
Comment (ngxson) [inline common/arg.h:92]: @ggerganov I'm not sure if it's a good idea to have a dedicated context for arg parser.

Before, `enum llama_example curr_ex` was a member of `gpt_params` so we don't have to create a new struct to hold it (same for `print_usage`)

However, when moving `enum llama_example` away from `common.h`, I got error `ISO C++ forbids forward references to 'enum' types` when trying to use it inside `common.h`

File: examples/batched-bench/batched-bench.cpp [hunk 2]
```diff
@@ -37,8 +38,8 @@ static void print_usage(int, char ** argv) {
 int main(int argc, char ** argv) {
     gpt_params params;
 
-    auto options = gpt_params_parser_init(params, LLAMA_EXAMPLE_BENCH, print_usage);
-    if (!gpt_params_parse(argc, argv, params, options)) {
+    auto ctx_arg = gpt_params_parser_init(params, LLAMA_EXAMPLE_BENCH, print_usage);
+    if (!gpt_params_parse(argc, argv, ctx_arg)) {
         return 1;
     }
```
Comment (ggerganov) [inline examples/batched-bench/batched-bench.cpp:44]: This repeating pattern can be simplified with a wrapper:

```suggestion
    if (!gpt_params_parse_ext(argc, argv, params, LLAMA_EXAMPLE_BENCH, print_usage)) {
        return 1;
    }
```

File: common/arg.h (hunk 1 above)
Comment (ggerganov) [reply common/arg.h:69]: I think it's OK to add it. `gpt_params_context` seems a better name though.

**Review by ggerganov — APPROVED** [review]

**Review by ggerganov — APPROVED** [review]

File: examples/batched-bench/batched-bench.cpp (hunk 2 above)
Comment (ngxson) [reply examples/batched-bench/batched-bench.cpp]: I ended up directly changing `gpt_params_parse` to `gpt_params_parse(argc, argv, params, LLAMA_EXAMPLE_BENCH, print_usage)`, so we have only one version of this function: https://github.com/ggerganov/llama.cpp/pull/9388/commits/96311e3248c9730b128b29999a4c8b7da0937ce9

File: common/arg.cpp [hunk 3]
```diff
@@ -0,0 +1,1994 @@
...
+#ifdef GGML_USE_SYCL
+                fprintf(stderr, "warning: The split mode value:[row] is not supported by llama.cpp with SYCL. It's developing.\nExit!\n");
+                exit(1);
+#endif // GGML_USE_SYCL
+                params.split_mode = LLAMA_SPLIT_MODE_ROW;
+            }
+            else {
+                throw std::invalid_argument("invalid value");
+            }
+#ifndef GGML_USE_CUDA_SYCL_VULKAN
```
Comment (m18coppola) [inline common/arg.cpp:1436]: I don't think this is working as intended

Comment (ngxson) [reply common/arg.cpp:1436]: can you suggest a patch?

Comment (ngxson) [reply common/arg.cpp:1436]: Yes we can move it here. However, the naming is quite misleading, so I'd suggest **not** to abstract it:

```suggestion
#if (!defined(GGML_USE_CUDA) && !defined(GGML_USE_SYCL)) && !defined(GGML_USE_VULKAN)
```

**Pull request merged**